File: code/fig3_roc.py
#!/usr/bin/env python3
"""Export ROC-curve data for Figure 3 (rendered by fig3_roc.R).
Writes roc.csv (interpolated to 500 points/label) to results_full/figs/."""
import os, csv, numpy as np
from sklearn.metrics import roc_curve, roc_auc_score
RES=os.path.join(os.environ.get("ECHONEXT_DATA", os.path.expanduser("~/Desktop/RESEARCH")), "EchoNext-repo/results_full"); OUT=os.path.join(RES,"figs"); os.makedirs(OUT,exist_ok=True)
probs=np.load(f"{RES}/probs.npy"); paths=[p for p in open(f"{RES}/kept_paths.txt").read().split("\n") if p]
coh={r['ecg_path']:r for r in csv.DictReader(open(os.path.join(os.environ.get("ECHONEXT_DATA", os.path.expanduser("~/Desktop/RESEARCH")), "cohort_oneperpt_full.csv")))}
COLS=['lvef_lte_45','lvwt_gte_13','aortic_stenosis_modsev','aortic_regurg_modsev','mitral_regurg_modsev','tricuspid_regurg_modsev','pulm_regurg_modsev','rv_dysfunction_modsev','pericardial_modlarge','pasp_gte_45','tr_max_gte_32','shd']
rows=[coh[p] for p in paths]; y=np.array([[int(r[c]) for c in COLS] for r in rows])
grid=np.linspace(0,1,500)
with open(f"{OUT}/roc.csv","w",newline="") as f:
    w=csv.writer(f); w.writerow(["label","fpr","tpr"])
    for j,name in [(11,'Any SHD'),(0,'Reduced LVEF'),(7,'RV dysfunction')]:
        fpr,tpr,_=roc_curve(y[:,j],probs[:,j]); auc=roc_auc_score(y[:,j],probs[:,j])
        tg=np.interp(grid,fpr,tpr); lab=f"{name} (AUROC {auc:.2f})"
        [w.writerow([lab,a,b]) for a,b in zip(grid,tg)]
print("exported roc.csv ->", OUT)

File: code/fig3_roc.R
#!/usr/bin/env Rscript
# Figure 3 — ROC curves (composite + strongest components) in MIMIC-IV external validation.
# Run: python3 code/fig3_roc.py  &&  Rscript code/fig3_roc.R
suppressMessages(library(ggplot2))
navy<-"#1f3b5c"; teal<-"#2a9d8f"; rust<-"#bb4430"
figd<-file.path(Sys.getenv("ECHONEXT_DATA", path.expand("~/Desktop/RESEARCH")), "EchoNext-repo/results_full/figs")
repo<-file.path(Sys.getenv("ECHONEXT_DATA", path.expand("~/Desktop/RESEARCH")), "echonext-mimic-validation/results/figures")
roc<-read.csv(file.path(figd,"roc.csv")); labs<-unique(roc$label); roc$label<-factor(roc$label, levels=labs)
cols<-setNames(c(navy,teal,rust), labs)   # Any SHD / Reduced LVEF / RV dysfunction (export order)
p<-ggplot(roc, aes(fpr, tpr, color=label))+
  geom_abline(slope=1,intercept=0,linetype="dashed",color="grey55",linewidth=0.4)+
  geom_line(linewidth=1)+ scale_color_manual(values=cols)+
  coord_equal(xlim=c(0,1),ylim=c(0,1),expand=FALSE)+
  labs(x="1 - Specificity", y="Sensitivity", title="ROC - MIMIC-IV external validation", color=NULL)+
  theme_classic(base_size=11, base_family="Helvetica")+
  theme(legend.position=c(0.97,0.04), legend.justification=c(1,0), legend.background=element_blank(),
        legend.key.height=unit(13,"pt"), plot.title=element_text(size=11), plot.margin=margin(8,10,6,6))
for (d in c(repo, figd)) { if(!dir.exists(d)) dir.create(d,recursive=TRUE)
  ggsave(file.path(d,"fig5_roc.png"), p, width=5.4, height=4.9, dpi=300, bg="white")
  ggsave(file.path(d,"fig5_roc.pdf"), p, width=5.4, height=4.9, bg="white") }
cat("saved fig5_roc (R/ggplot2)\n")
